datalad/interface/rerun.py

```diff
@@ -0,0 +1,300 @@
+            metavar="base",
+            args=("--onto",),
+            doc="""start point for rerunning the commands. If not
+            specified, commands are executed at HEAD. This option can
+            be used to specify an alternative start point, which will
+            be checked out with the branch name specified by --branch
+            or in a detached state otherwise. As a special case, an
+            empty value for this option means to use the commit
+            specified by --since.""",
+            constraints=EnsureStr() | EnsureNone()),
+        # TODO
+        # --list-commands
+        #   go through the history and report any recorded command. this info
+        #   could be used to unlock the associated output files for a rerun
+    )
+
+    @staticmethod
+    @datasetmethod(name='rerun')
+    @eval_results
+    def __call__(
+            revision="HEAD",
+            since=None,
+            dataset=None,
+            branch=None,
+            message=None,
+            onto=None):
+
+        ds = require_dataset(
+            dataset, check_installed=True,
+            purpose='rerunning a command')
+
+        lgr.debug('rerunning command output underneath %s', ds)
+
+        from datalad.tests.utils import ok_clean_git
+        try:
+            ok_clean_git(ds.path)
+        except AssertionError:
+            yield get_status_dict(
+                'run',
+                ds=ds,
+                status='impossible',
+                message=('unsaved modifications present, '
+                         'cannot detect changes by command'))
+            return
+
+        err_info = get_status_dict('run', ds=ds)
+        if not ds.repo.get_hexsha():
+            yield dict(
+                err_info, status='impossible',
+                message='cannot rerun command, nothing recorded')
+            return
+
+        if branch and branch in ds.repo.get_branches():
+            yield get_status_dict(
+                "run", ds=ds, status="error",
+                message="branch '{}' already exists".format(branch))
+            return
+
+        root = False
+        if since is None:
+            revrange = "{}^..{}".format(revision, revision)
+        elif since.strip() == "":
+            revrange = revision
+            root = True
+        else:
+            revrange = "{}..{}".format(since, revision)
+
+        if ds.repo.repo.git.rev_list("--merges", revrange, "--"):
+            yield get_status_dict(
+                "run", ds=ds, status="error",
+                message="cannot rerun history with merge commits")
+            return
+        revs = ds.repo.repo.git.rev_list("--reverse", revrange, "--").split()
+
+        do_checkout = branch
+        if onto is not None:
+            if onto.strip() == "":
+                ## An empty argument means go to the parent of the
+                ## first revision, but that doesn't exist for
+                ## --since=.  Instead check out an orphan branch.
```

Comment:
use of `--onto=` generated a new branch not rooted on the first `RUN` commit's parent or some other first commit where things actually diverged, which I thought would be the goal here.
my command was `datalad rerun --since= -b rerun2 --onto=`

here is a full list of commands in history
```
datalad create /tmp/test111
cd /tmp/test111
datalad run "echo 123 > 123"
vim 111 # added blah
git add 111
git commit -m 'added 111'
datalad rerun --since=
datalad run "date > date"
vim 111 # added blah
git commit -m 'more blah' 111
datalad rerun --since=
datalad rerun --since= --onto=
datalad rerun --since= -b rerun1
git co master
datalad rerun --since= -b rerun2 --onto=
```

Reply:
> which I thought would be the goal here

No, that wasn't my goal :]  It was a simple "parent of the first
commit in the specified range".  And if the root of the branch is
specified, this would mean an orphan branch.  I haven't thought much
about starting at the first run command, so I'll look into it, but I
do think I prefer the conceptual simplicity of just using the
specified range.